src/main.c

```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
   hd_dbus_send_event (s);
 }
 
+/* Close all input devices previously opened */
+static void
+close_input_devices (Display *dpy)
+{
+  int i;
+
+  for (i = 0; i < xi_devices->len; i++)
+    {
+      hd_xi_device *xi_dev = &g_array_index (xi_devices, hd_xi_device, i);
+
+      if (xi_dev->dev)
+        {
+          XCloseDevice (dpy, xi_dev->dev);
+          xi_dev->dev = NULL;
+        }
+    }
+}
+
+/* Fill in the array with input devices type is touchscreen */
+static void
+enumerate_input_devices (Display *dpy)
+{
+  static Atom atom_touchscreen = None;
+  XDeviceInfo *devinfo;
+  int i, ndev;
+  GArray *eclass;
+
+  if (xi_devices)
+    {
+      close_input_devices (dpy);
+      g_array_free (xi_devices, TRUE);
+      xi_devices = NULL;
+    }
```

Comment:
I'd move this logic to `close_input_devices()` which may be renamed to `free_input_devices`.
So we just call `free_input_devices()` here and on hildon-desktop termination.

Reply:
Yes, it can be done as you say, but it is a matter of preference, There is no gain.

Also, freeing function a() allocated memory in function b() is bad coding practice, that's why it is done like that.